src/radiation/mocmc.cpp

```diff
@@ -0,0 +1,1100 @@
+      nblock - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, int &nsamp) {
+        Real Jtot = 0.;
+        for (int s = 0; s < 3; s++) {
+          Jtot += v(b, pJ(s), k, j, i);
+        }
+        v(b, dn, k, j, i) =
+            get_nsamp_per_zone(k, j, i, geom, v(b, pdens, k, j, i), v(b, pT, k, j, i),
+                               v(b, pye, k, j, i), Jtot, nsamp_per_zone);
+        nsamp += v(b, dn, k, j, i);
+      },
+      Kokkos::Sum<int>(nsamp_tot));
+
+  ParArrayND<int> new_indices;
+  auto new_mask = swarm->AddEmptyParticles(nsamp_tot, new_indices);
+
+  // Calculate array of starting index for each zone to compute particles
+  ParArrayND<int> starting_index("Starting index", nx_k, nx_j, nx_i);
+  auto starting_index_h = starting_index.GetHostMirror();
+  auto dN = rc->Get(im::dnsamp).data;
+  auto dN_h = dN.GetHostMirrorAndCopy();
+  int index = 0;
+  for (int k = 0; k < nx_k; k++) {
+    for (int j = 0; j < nx_j; j++) {
+      for (int i = 0; i < nx_i; i++) {
+        starting_index(k, j, i) = index;
+        index += static_cast<int>(dN_h(k + kb.s, j + jb.s, i + ib.s));
+      }
+    }
+  }
+  starting_index.DeepCopy(starting_index_h);
+
+  const auto &x = swarm->template Get<Real>("x").Get();
+  const auto &y = swarm->template Get<Real>("y").Get();
+  const auto &z = swarm->template Get<Real>("z").Get();
+  const auto &ncov = swarm->template Get<Real>("ncov").Get();
+  const auto &Inuinv = swarm->template Get<Real>("Inuinv").Get();
+
+  auto swarm_d = swarm->GetDeviceContext();
+
+  auto nusamp = rad->Param<ParArray1D<Real>>("nusamp");
+  const int nu_bins = rad->Param<int>("nu_bins");
+  auto species = rad->Param<std::vector<RadiationType>>("species");
+  auto num_species = rad->Param<int>("num_species");
+  RadiationType species_d[3] = {};
+  for (int s = 0; s < num_species; s++) {
+    species_d[s] = species[s];
+  }
+
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "MOCMC::Init::Sample", DevExecSpace(), 0, nblock - 1, kb.s,
+      kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        const int start_idx = starting_index(k - kb.s, j - jb.s, i - ib.s);
+        auto rng_gen = rng_pool.get_state();
+
+        Real cov_g[4][4];
+        geom.SpacetimeMetric(CellLocation::Cent, k, j, i, cov_g);
+        Real alpha = geom.Lapse(CellLocation::Cent, k, j, i);
+        Real con_beta[3];
+        geom.ContravariantShift(CellLocation::Cent, k, j, i, con_beta);
+        const Real vpcon[] = {v(pv(0), k, j, i), v(pv(1), k, j, i), v(pv(2), k, j, i)};
+        const Real W = phoebus::GetLorentzFactor(vpcon, cov_g);
+        const Real ucon[4] = {W / alpha, vpcon[0] - con_beta[0] * W / alpha,
+                              vpcon[1] - con_beta[1] * W / alpha,
+                              vpcon[2] - con_beta[2] * W / alpha};
```

Comment:
wonder if this should be a relativity utils function

Reply:
Yeah I should move it there :+1:

Reply:
NVM `GetFourVelocity` is already in relativity utils :)

Reply:
I guess I'm starting to worry about increasing the complexity of relativity utils any further... to avoid the situation where you end up calling a function that calculates gcov and then calls another subfunction that also calculates gcov for example.

Reply:
yeah that's a concern we should be aware of. Not sure what the solution is except try to design things with the ability to pass in other needed quantities. idk

Reply:
Yeah, or like receive a struct that checks at compile time what geometric and physical quantities have already been calculated and uses `constexpr if` to calculate anything needed but not yet known. Sounds pretty complicated.

Reply:
That sounds way too complicated, lol. For now I think we punt. But something to be aware of.